src/Microsoft.AspNet.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -217,7 +217,10 @@ public void OnCompleted(Action continuation)
             else
             {
                 // THIS IS AN ERROR STATE - ONLY ONE WAITER CAN WAIT
-                throw new InvalidOperationException("Concurrent reads are not supported.");
```

Comment:
Also think we should make the error bubble to app code instead of throwing on some background thread. Set the error and trigger the callback.